CanonicalABI.md

```diff
@@ -0,0 +1,976 @@
+`pull-buffer` are represented with a singular `i32` instead of three `i32`
+values. This is because if an export is receiving a buffer it's coming from
+somewhere else which means it's a handle to a buffer and various "intrinsics"
+will be used to operate on the handle.
+
+Some other helper-like functions used above and below are:
+
+```
+# Used to determine whether a record is represented as bitflags or not.
+def is_bitflags(fields):
+    fields.all(|f| f == bool) && fields.len() <= 64
+
+```
+
+## Lifting values
+
+With the core wasm signature above, the definition of how to lift values in
+import/export contexts then looks like:
+
+```
+def lift_all(direction, types, values):
+    values = values.iter()
+    types.map(|ty| lift(direction, ty, values))
+
+def lift(direction, type, values):
+  # assume that `assert_next(ty)` returns a wasm value of type
+  # `ty` after asserting the value is indeed of type `ty`
+  match type
+    f32 => values.assert_next(f32) as f32
+    f64 => values.assert_next(f64) as f64
+
+    # These conversions assume no truncation happens.
+    s8 => values.assert_next(i32) as s8
+    u8 => values.assert_next(i32) as u8
+    s16 => values.assert_next(i32) as s16
+    u16 => values.assert_next(i32) as u16
+
+    s32 => values.assert_next(i32) as s32
+    u32 => values.assert_next(i32) as u32
+    s64 => values.assert_next(i64) as s64
+    u64 => values.assert_next(i64) as u64
+
+    # This assumes that the i32 is a valid unicode code point.
+    char => values.assert_next(i32) as char
+
+    list $ty => {
+        ptr = values.assert_next(i32)
+        len = values.assert_next(i32)
+
+        # Lists returned in imports are allocated resources that are now
+        # owned by the current module which means that we need to call a
+        # free function of some form. For now that function is called
+        # "witx_free" and takes pointer/size/alignment arguments
+        if direction == "import":
+          defer module.witx_free(ptr, len * sizeof(direction, $ty), align(direction, $ty))
+
+        if $ty == char:
+          # note that this can skip validation due to the canonical ABI
+          read_utf8_string(ptr, len)
+        else:
+          # note that this can be optimized to a `memcpy` in some cases, and
+          # otherwise `read_from_memory` is defined below.
+          result = []
+          for i in 0..len:
+              read_from_memory(direction, $ty, ptr)
+              ptr += size(direction, $ty)
+          result
+    }
+
+    record $fields => {
+        # If this record is represented with bitflags then we pop the
+        # singular value and wrap it up in the language's representation of
+        # what bitflags are. Note that the bitflags bits are assumed valid
+        # here.
+        if is_bitflags($fields):
+          if $fields.len() > 32:
+              values.assert_next(i64) as $bitflags<$fields>
+          else:
+              values.assert_next(i32) as $bitflags<$fields>
```

Comment:
Here's another interesting argument for treating `(flags "a" "b")` different from `(record (field "a" bool) (field "b" bool))`: if we're simply optimizing the canonical ABI of certain records, then the produced language value should be a `$language_struct` (as below).  But if we're recognizing the `flags` abbreviation, then we could produce a language-visibly-different value (e.g., an actual bitvector).

Reply:
This is related to the `string` thing above where I've been working under the assumption that the syntactic sugar isn't present in the AST. In that sense the two records you described here wouldn't have a different representation in the source language, they'd always be bitflags.

I do think it's somewhat nice to not have too much special-casing and too complicated of an AST here for the type grammar, and I'm not sure if there's much use in having different canonical ABI representations for the same structs?